File: 题库/[144]二叉树的前序遍历.c
//给定一个二叉树，返回它的 前序 遍历。 
//
// 示例: 
//
// 输入: [1,null,2,3]  
//   1
//    \
//     2
//    /
//   3 
//
//输出: [1,2,3]
// 
//
// 进阶: 递归算法很简单，你可以通过迭代算法完成吗？ 
// Related Topics 栈 树

// 前序遍历思想
// 根节点->左节点->右节点


//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     struct TreeNode *left;
 *     struct TreeNode *right;
 * };
 */

/**
 * 获取目标树的节点个数
 * @param root
 * @return
 */
int getTreeSize(struct TreeNode* root) {
    if (!root) return 0;
    return getTreeSize(root->left) + getTreeSize(root->right) + 1;
}

/**
 * 递归前序遍历二叉树并保存输出值和数组长度
 * @param root
 * @param res
 * @param returnSize
 */
void preOrder(struct TreeNode* root, int *res, int *returnSize) {
    if (!root) return;
    res[(*returnSize)++] = root->val;
    preOrder(root->left, res, returnSize);
    preOrder(root->right, res, returnSize);
}

/**
 * * 二叉树的前序遍历递归算法
 * Note: The returned array must be malloced, assume caller calls free().
 * @param root : 树的根节点
 * @param returnSize : 返回数组的长度用来在递归过程中记录当前数组的下标
 * @return
 */
int* preorderTraversal(struct TreeNode* root, int* returnSize){
    if (!root) {
        *returnSize = 0;
        return NULL;
    }

    int treeSize = getTreeSize(root);
    int *res = malloc(sizeof(int) * treeSize);
    *returnSize = 0;

    preOrder(root, res, returnSize);
    return res;
}


//leetcode submit region end(Prohibit modification and deletion)

File: tree/leetcode/editor/cn/[144]二叉树的前序遍历.go
//给你二叉树的根节点 root ，返回它节点值的 前序 遍历。 
//
// 
//
// 示例 1： 
//
// 
//输入：root = [1,null,2,3]
//输出：[1,2,3]
// 
//
// 示例 2： 
//
// 
//输入：root = []
//输出：[]
// 
//
// 示例 3： 
//
// 
//输入：root = [1]
//输出：[1]
// 
//
// 示例 4： 
//
// 
//输入：root = [1,2]
//输出：[1,2]
// 
//
// 示例 5： 
//
// 
//输入：root = [1,null,2]
//输出：[1,2]
// 
//
// 
//
// 提示： 
//
// 
// 树中节点数目在范围 [0, 100] 内 
// -100 <= Node.val <= 100 
// 
//
// 
//
// 进阶：递归算法很简单，你可以通过迭代算法完成吗？ 
// Related Topics 栈 树 
// 👍 553 👎 0

// 前序遍历：根->左->右
// 中序遍历：左->根->右
// 后续遍历：左->右->根

// 迭代写法思想：
// 1. 借助栈
// 2. 无论前中后遍历，都是先遍历左子树直到叶子节点（入栈），然后在一个节点一个节点出栈做处理
// 2.1：前序遍历要优先遍历根节点，所以在入栈时，就打印
// 2.2：中遍历要优先遍历左节点，所以在出栈时，打印根节点
// 2.3：后续遍历左右根，所以需要记录上一次遍历的节点是什么，才能找到对应根节点，
// 入栈，获取栈顶元素，判断有没有右节点，有则继续入栈，没有则判断当前节点是叶子节点还是根节点（node->right == lastNode）
// 3. 处理右子树

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func preorderTraversal(root *TreeNode) []int {
	if root == nil {
		return nil
	}

	result := make([]int, 0)
	stack := make([]*TreeNode, 0)
	for root != nil || len(stack) > 0 {
		for root != nil {
			result = append(result, root.Val)
			stack = append(stack, root)
			root = root.Left
		}

		node := stack[len(stack) - 1]
		stack = stack[:len(stack) - 1]

		root = node.Right
    }

    return result
}
//leetcode submit region end(Prohibit modification and deletion)
